### PR TITLE
Add the recording of events

### PR DESCRIPTION
**Pull request by rihafilip**

### CODE REVIEW

**Comment by o-** [comment]
Here are some stats for your PR:

* WARNING: the longest CI job test_release_2 took 2.43h
* binarytrees_2 regressed by 0.95
* Overall benchmarks improved by 1.0

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=6939281555&job_ids[]=6886467222&selection=all

**Comment by o-** [comment]
Here are some stats for your PR:

* WARNING: the longest CI job test_release_2 took 2.65h
* scalarFor regressed by 0.78
* Overall benchmarks improved by 1.0

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=6959319162&job_ids[]=6941702526&selection=all

File: CMakeLists.txt [hunk 1]
```diff
@@ -32,6 +32,12 @@ if(${MACOS_USE_GCC_9})
     set(CMAKE_CXX_COMPILER /usr/local/bin/g++-9 CACHE PATH "" FORCE)
 endif()
 
+option(RECORDING_HOOKS "Enable recording hooks, which record specific events" FALSE)
```
Comment (fikovnik) [inline CMakeLists.txt]: Lets call it just RECORDING, RECORDING_HOOKS seem like an impl detail?

File: replayer.r [hunk 2]
```diff
@@ -0,0 +1,133 @@
+recordings.csv <- function( r ) {
```
Comment (fikovnik) [inline replayer.r]: Is that your code formatting - adding the extra space?
Would you mind changing it so the R code in the repo is somewhat similar...

File: rir/R/rir.R [hunk 3]
```diff
@@ -216,3 +216,56 @@ rir.annotateDepromised <- function(closure) {
     rir.markFunction(copy, DepromiseArgs=TRUE)
     copy
 }
+
+# Only define recording functions if the recordings have been compiled
+if ( .Call("isRecordingsDefined") ){
```
Comment (fikovnik) [inline rir/R/rir.R]: extra spaces

Comment (fikovnik) [reply rir/R/rir.R]: and missing space :)

File: rir/src/interpreter/interp.cpp [hunk 4]
```diff
@@ -1006,12 +1008,16 @@ SEXP doCall(CallContext& call, bool popArgs) {
                         call.caller->size() < pir::Parameter::MAX_INPUT_SIZE &&
                         fun->body()->codeSize <
                             pir::Parameter::PIR_OPT_BC_SIZE) {
+
+                        REC_HOOK(recording::recordOsrTriggerCallerCalle());
```
Comment (fikovnik) [inline rir/src/interpreter/interp.cpp]:
```suggestion
                        REC_HOOK(recording::recordOsrTriggerCallerCallee());
```

File: rir/src/interpreter/interp.h [hunk 5]
```diff
@@ -59,42 +60,65 @@ inline bool RecompileHeuristic(Function* fun,
...
     if (flags.contains(Function::NotOptimizable))
         return false;
 
     if (!funMaybeDisabled)
         funMaybeDisabled = fun;
 
-    auto abandon =
-        funMaybeDisabled->deoptCount() >= pir::Parameter::DEOPT_ABANDON;
+    auto abandon
+        = funMaybeDisabled->deoptCount() >= pir::Parameter::DEOPT_ABANDON;
```
Comment (fikovnik) [inline rir/src/interpreter/interp.h]:
```suggestion
    auto abandon = 
            funMaybeDisabled->deoptCount() >= pir::Parameter::DEOPT_ABANDON;
```

Comment (fikovnik) [reply rir/src/interpreter/interp.h]: The less changes the better

File: rir/src/interpreter/interp.h [hunk 6]
```diff
@@ -59,42 +60,65 @@ inline bool RecompileHeuristic(Function* fun,
...
+        return true;
+    }
+
+    if (!fun->isOptimized()) {
+        REC_HOOK(recording::recordNotOptimizedReason());
+        return true;
+    }
+
+    if (context.smaller(fun->context()) &&
+        context.isImproving(fun) > table->size()){
```
Comment (fikovnik) [inline rir/src/interpreter/interp.h]:
```suggestion
        context.isImproving(fun) > table->size()) {
```

File: rir/src/prefixer.cpp
Comment (fikovnik) [inline rir/src/prefixer.cpp:1]: Does this need to be a separate file? Is it used from many places?

File: rir/src/runtime/RirRuntimeObject.h [hunk 7]
```diff
@@ -74,6 +74,7 @@ struct RirRuntimeObject {
     RirRuntimeObject(uint32_t gc_area_start, uint32_t gc_area_length)
         : info{gc_area_start, gc_area_length, MAGIC} {
         uint8_t* start = (uint8_t*)this + gc_area_start;
+        // cppcheck-suppress pointerSize
```
Comment (fikovnik) [inline rir/src/runtime/RirRuntimeObject.h:77]: Yeah, this is indeed weird

File: rir/tests/recording_basics.R
Comment (fikovnik) [inline rir/tests/recording_basics.R:1]: Some tests would be good :-)

**Review by fikovnik — APPROVED** [review]
A few minor things.

Later we should look at some testing.

File: rir/src/runtime/RirRuntimeObject.h (hunk 7 above)
Comment (rihafilip) [reply rir/src/runtime/RirRuntimeObject.h:77]: The commit where it breaks is this one https://github.com/reactorlabs/rir/pull/1282/commits/81f05b91284ceecb467673cb3e97f566bd135497, which only moves code around, so it is weird that CppChecks triggers only after that.

Comment (rihafilip) [reply rir/src/runtime/RirRuntimeObject.h:77]: The same error is raised even on the newest CppCheck version (2.14.1).

**Comment by o-** [comment]
Here are some stats for your PR:

* WARNING: the longest CI job test_release_2 took 2.44h
* real_thing suite improved by 1.03
* reversecomplement_2 regressed by 0.94
* scalarFor regressed by 0.87
* Overall benchmarks improved by 1.0

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=6980617871&job_ids[]=6941702526&selection=all

**Pull request merged**